src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpStreamValidation.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Xml;
+using System.Xml.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using NuGet.Packaging;
+
+namespace NuGet.Protocol.Core.v3
+{
+    public static class HttpStreamValidation
+    {
+        public static void ValidateJObject(string uri, Stream stream, Action<JObject> validate)
+        {
+            try
+            {
+                using (var reader = new StreamReader(stream, Encoding.UTF8, false, 4096, true))
+                using (var jsonReader = new JsonTextReader(reader) { CloseInput = false })
+                {
+                    var jObject = JObject.Load(jsonReader);
+                    validate(jObject);
+                }
+            }
+            catch (Exception e) when (!(e is InvalidDataException))
+            {
+                string message = string.Format(
+                    CultureInfo.CurrentCulture,
+                    Strings.Protocol_InvalidJsonObject,
+                    uri);
+
+                throw new InvalidDataException(message, e);
+            }
+        }
+        public static void ValidateJObject(string uri, Stream stream)
+        {
+            ValidateJObject(uri, stream, o => { });
+        }
+
+        public static void ValidateNupkg(string uri, Stream stream)
+        {
+            try
+            {
+                using (var reader = new PackageArchiveReader(stream, leaveStreamOpen: true))
+                using (reader.GetNuspec())
```

Comment:
This can return null and the using would handle it. It might be better to just open this as a ZipArchive and check that it has entries.